Slim/Formats/Movie.pm

```diff
@@ -200,4 +206,203 @@ sub findFrameBoundaries {
 
 sub canSeek { 1 }
 
+sub parseStream {
+	my ( $class, $dataref, $args, $formats ) = @_;
+	return -1 unless defined $$dataref;
+	
+	# stitch new data to existing buf and init parser if needed
+	$args->{_scanbuf} .= $$dataref;
+	$args->{_need} ||= 8;
+	$args->{_offset} ||= 0;
+	
+	my $len = length($$dataref);
+	my $offset = $args->{_offset};
+	my $log = logger('player.streaming');
+	
+	while (length($args->{_scanbuf}) > $args->{_offset} + $args->{_need} + 8) {
+		$args->{_atom} = substr($args->{_scanbuf}, $offset+4, 4);
+		$args->{_need} = unpack('N', substr($args->{_scanbuf}, $offset, 4));
+		$args->{_offset} = $args->{"_$args->{_atom}_"} = $offset;
+		
+		# a bit of sanity check
+		if ($offset == 0 && $args->{_atom} ne 'ftyp') {
+			$log->warn("no header! this is supposed to be a mp4 track");
+			return 0;
+		}
+		
+		$offset += $args->{_need};
+		main::DEBUGLOG && $log->is_debug && $log->debug("atom $args->{_atom} at $args->{_offset} of size $args->{_need}");
+		
+		# mdat reached = audio offset & size acquired
+		if ($args->{_atom} eq 'mdat') {
+			$args->{_audio_size} = $args->{_need};
+			last;
+		}
+	}
+	
+	return -1 unless $args->{_mdat_};
+
+	# now make sure we have acquired a full moov atom
+	if (!$args->{_moov_}) {
+		# no 'moov' found but EoF
+		if (!$len) {
+			$log->warn("no 'moov' found before EOF => track probably not playable");
+			return 0;
+		}
+		
+		# already waiting for bottom 'moov', we need more
+		return -1 if $args->{_range};
+		
+		# top 'moov' not found, need to seek beyond 'mdat'
+		$args->{_range} = $offset;
+		$args->{_scanbuf} = substr($args->{_scanbuf}, 0, $args->{_offset});
+		delete $args->{_need};
+		return $offset;
+	} elsif ($args->{_atom} eq 'moov' && $len) {
+		return -1;
+	}	
+	
+	# finally got it, add 'moov' size it if was last atom
+	$args->{_scanbuf} = substr($args->{_scanbuf}, 0, $args->{_offset} + ($args->{_atom} eq 'moov' ? $args->{_need} : 0));
+	
+	# put at least 16 bytes after mdat or it confuses audio::scan (and header creation)
+	my $fh = File::Temp->new();
+	$fh->write($args->{_scanbuf} . pack('N', $args->{_audio_size}) . 'mdat' . ' ' x 16);
+	$fh->seek(0, 0);
+
+	my $info = Audio::Scan->scan_fh( mp4 => $fh )->{info};
+	$info->{fh} = $fh;
```

Comment:
I think we could use `Slim::Player::Song::DESTROY` to close the handle. Something like this:

```diff
diff --git a/Slim/Player/Song.pm b/Slim/Player/Song.pm
index 5db48cf7f..d965607f4 100644
--- a/Slim/Player/Song.pm
+++ b/Slim/Player/Song.pm
@@ -156,6 +156,12 @@ sub new {
 
 sub DESTROY {
        my $self = shift;
+
+       my $track = $self->_track;
+       my $fh = $track->initial_block_fh if $track->can('initial_block_fh');
+       $track->initial_block_fh(undef);
+       close $fh if $fh;
+
        $_liveCount--;
        if (main::DEBUGLOG && $log->is_debug)   {
                $log->debug(sprintf("DESTROY($self) live=$_liveCount: index=%d, url=%s", $self->index(), $self->_track()->url));
```

This effectively got rid of the temporary files. `close $track->initial_block_fh` did not work, for whatever reason.

Reply:
Do you think RemoteTrack::delete could be used instead?

Reply:
I've also tried to add a DESTROY method to RemoteTrack but it does not seem to be called. I have to check the Song as well but I think I remember in a previous modification, when I was trying to understand how all this works, I tested this liveCount and I never saw it decreased

Reply:
`RemoteTrack::delete()` is more than that. I actually was wondering whether that information/file handle should belong to the song object rather than the track, as it's only relevant when the track is being played. The track object can be around for much longer, eg. when it's queued up, or after it had been played, it would still be part of the current playlist. But the file handle is not required at that point.

Reply:
But I think this fh is really an attribute of the track that must 'survive' between $song instances as the url might not be rescanned. So if I delete it, I won't be able to re-create it. And when I populate the track information, I might not have a song object. For example, scanUrl does not always have one when called, as it can scan multiple tracks, keep them ready and then create song when getNextTrack/Song is called. It's like for File.pm, the audio_offset and audio_size (e.g.) are attributes that are created independently of a song object. But I agree that we don't want to keep them around forever. I'm trying to find under which circumstances the $track object is destroyed (other than the call to delete which obviously happen when the track changes)

Reply:
In order to see whether `RemoteTrack`'s `DESTROY` would be called or not you could reduce the size of `%Cache` in that file. It's initially set to 500 items, but in `init()`'s `$maxPlaylistLengthCB` you could set it to any small size to make sure it soon would need to run some cleanup.

And yes, I've seen LMS clean up behind, too. But some people listen to hundreds of tracks before they restart LMS.

Reply:
ok, I will try that

Reply:
And I did not know it is possible to register a callback on a prefs change ... there are so many well thought features in LMS. I've discovered a couple of inconsistencies in codec management between pcm/wav/aif and mp4/aac but these are very minor compared to extremely well done global architecture, IMHO